File: src/store/modules/trade.ts
import { defineStore } from 'pinia';
import { ref, computed } from 'vue';
import type { Order, Position, OrderSide, OrderType, OrderStatus } from '@/types';
import * as api from '@/api/tauri';

/**
 * Place Order Request Interface
 */
export interface PlaceOrderRequest {
  symbol: string;
  side: OrderSide;
  orderType: OrderType;
  quantity: number;
  price?: number;
  stopPrice?: number;
  clientOrderId?: string;
  timeInForce?: 'GTC' | 'IOC' | 'FOK';
}

/**
 * Trade Store - Manages trading state and operations
 */
export const useTradeStore = defineStore('trade', () => {
  // ========== State ==========
  const orders = ref<Order[]>([]);
  const positions = ref<Position[]>([]);
  const trades = ref<any[]>([]); // Trade history
  const loading = ref(false);
  const error = ref<string | null>(null);

  // ========== Computed Getters ==========

  /**
   * Get active orders (open or partially filled)
   */
  const activeOrders = computed(() =>
    orders.value.filter(
      o => o.status === 'open' || o.status === 'partially_filled'
    )
  );

  /**
   * Get completed orders (filled, canceled, or rejected)
   */
  const completedOrders = computed(() =>
    orders.value.filter(
      o => o.status === 'filled' || o.status === 'canceled' || o.status === 'rejected'
    )
  );

  /**
   * Get orders by symbol
   */
  const getOrdersBySymbol = computed(() => {
    return (symbol: string) => orders.value.filter(o => o.symbol === symbol);
  });

  /**
   * Get positions by symbol
   */
  const getPositionsBySymbol = computed(() => {
    return (symbol: string) => positions.value.filter(p => p.symbol === symbol);
  });

  /**
   * Calculate total unrealized PnL
   */
  const totalUnrealizedPnl = computed(() =>
    positions.value.reduce((sum, p) => sum + p.unrealizedPnl, 0)
  );

  /**
   * Calculate total realized PnL
   */
  const totalRealizedPnl = computed(() =>
    positions.value.reduce((sum, p) => sum + p.realizedPnl, 0)
  );

  /**
   * Get long positions
   */
  const longPositions = computed(() =>
    positions.value.filter(p => p.side === 'long')
  );

  /**
   * Get short positions
   */
  const shortPositions = computed(() =>
    positions.value.filter(p => p.side === 'short')
  );

  // ========== Actions ==========

  /**
   * Fetch orders from backend
   */
  async function fetchOrders(userId: string, symbol?: string, status?: OrderStatus) {
    try {
      loading.value = true;
      error.value = null;

      const params: any = { userId };
      if (symbol) params.symbol = symbol;
      if (status) params.status = status;

      const data = await api.invokeRaw<Order[]>('trade_get_orders', params);
      orders.value = data;

      return data;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch orders';
      console.error('Failed to fetch orders:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Fetch open orders
   */
  async function fetchOpenOrders(userId: string) {
    try {
      loading.value = true;
      error.value = null;

      const data = await api.invokeRaw<Order[]>('trade_get_open_orders', { userId });
      orders.value = data;

      return data;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch open orders';
      console.error('Failed to fetch open orders:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Fetch positions from backend
   */
  async function fetchPositions(userId: string) {
    try {
      loading.value = true;
      error.value = null;

      const data = await api.invokeRaw<Position[]>('trade_get_positions', { userId });
      positions.value = data;

      return data;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to fetch positions';
      console.error('Failed to fetch positions:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Place a new order
   */
  async function placeOrder(userId: string, request: PlaceOrderRequest) {
    try {
      loading.value = true;
      error.value = null;

      // Convert to backend format
      const orderRequest = {
        symbol: request.symbol,
        side: request.side,
        orderType: request.orderType,
        quantity: request.quantity,
        price: request.price,
        stopPrice: request.stopPrice,
        clientOrderId: request.clientOrderId,
        timeInForce: request.timeInForce,
      };

      const order = await api.invokeRaw<Order>('trade_place_order', {
        userId,
        request: orderRequest,
      });

      // Add to orders list
      orders.value.push(order);

      return order;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to place order';
      console.error('Failed to place order:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Cancel an order
   */
  async function cancelOrder(userId: string, orderId: string) {
    try {
      loading.value = true;
      error.value = null;

      await api.invokeRaw<void>('trade_cancel_order', { userId, orderId });

      // Update local order status
      const orderIndex = orders.value.findIndex(o => o.id === orderId);
      if (orderIndex >= 0) {
        orders.value[orderIndex].status = 'canceled' as OrderStatus;
      }

      return true;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to cancel order';
      console.error('Failed to cancel order:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Cancel all orders (optionally by symbol)
   */
  async function cancelAllOrders(userId: string, symbol?: string) {
    try {
      loading.value = true;
      error.value = null;

      const params: any = { userId };
      if (symbol) params.symbol = symbol;

      const count = await api.invokeRaw<number>('trade_cancel_all_orders', params);

      // Update local orders status
      if (symbol) {
        orders.value
          .filter(o => o.symbol === symbol && (o.status === 'open' || o.status === 'partially_filled'))
          .forEach(o => o.status = 'canceled' as OrderStatus);
      } else {
        orders.value
          .filter(o => o.status === 'open' || o.status === 'partially_filled')
          .forEach(o => o.status = 'canceled' as OrderStatus);
      }

      return count;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to cancel all orders';
      console.error('Failed to cancel all orders:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Sync order status from exchange
   */
  async function syncOrderStatus(userId: string, orderId: string) {
    try {
      loading.value = true;
      error.value = null;

      const updatedOrder = await api.invokeRaw<Order>('trade_sync_order_status', {
        userId,
        orderId,
      });

      // Update local order
      const index = orders.value.findIndex(o => o.id === orderId);
      if (index >= 0) {
        orders.value[index] = updatedOrder;
      }

      return updatedOrder;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to sync order status';
      console.error('Failed to sync order status:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Close a position
   */
  async function closePosition(userId: string, symbol: string, side: 'long' | 'short', quantity?: number) {
    try {
      loading.value = true;
      error.value = null;

      const pnl = await api.invokeRaw<number>('trade_close_position', {
        userId,
        symbol,
        side,
        quantity,
      });

      // Refresh positions after closing
      await fetchPositions(userId);

      return pnl;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to close position';
      console.error('Failed to close position:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Get account balance
   */
  async function getBalance() {
    try {
      loading.value = true;
      error.value = null;

      const balance = await api.invokeRaw<any[]>('trade_get_balance');
      return balance;
    } catch (err) {
      error.value = err instanceof Error ? err.message : 'Failed to get balance';
      console.error('Failed to get balance:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  /**
   * Update order locally (from WebSocket event)
   */
  function updateOrder(order: Order) {
    const index = orders.value.findIndex(o => o.id === order.id);
    if (index >= 0) {
      orders.value[index] = order;
    } else {
      orders.value.push(order);
    }
  }

  /**
   * Update position locally (from WebSocket event)
   */
  function updatePosition(position: Position) {
    const index = positions.value.findIndex(
      p => p.symbol === position.symbol && p.side === position.side
    );
    if (index >= 0) {
      positions.value[index] = position;
    } else {
      positions.value.push(position);
    }
  }

  /**
   * Add trade to history
   */
  function addTrade(trade: any) {
    trades.value.unshift(trade);
    // Keep max 1000 trades
    if (trades.value.length > 1000) {
      trades.value = trades.value.slice(0, 1000);
    }
  }

  /**
   * Clear all data
   */
  function clear() {
    orders.value = [];
    positions.value = [];
    trades.value = [];
    error.value = null;
  }

  /**
   * Initialize trade data
   */
  async function initialize(userId: string) {
    try {
      loading.value = true;
      await Promise.all([
        fetchOpenOrders(userId),
        fetchPositions(userId),
      ]);
    } catch (err) {
      console.error('Failed to initialize trade store:', err);
      throw err;
    } finally {
      loading.value = false;
    }
  }

  return {
    // State
    orders,
    positions,
    trades,
    loading,
    error,

    // Computed
    activeOrders,
    completedOrders,
    getOrdersBySymbol,
    getPositionsBySymbol,
    totalUnrealizedPnl,
    totalRealizedPnl,
    longPositions,
    shortPositions,

    // Actions
    fetchOrders,
    fetchOpenOrders,
    fetchPositions,
    placeOrder,
    cancelOrder,
    cancelAllOrders,
    syncOrderStatus,
    closePosition,
    getBalance,
    updateOrder,
    updatePosition,
    addTrade,
    clear,
    initialize,
  };
});

File: src-tauri/src/commands/trade.rs
//! Trade commands for Tauri
//!
//! This module provides Tauri command handlers for trading operations.

use crate::core::response::{ApiResponse, ApiError};
use crate::core::trade::types::*;
use crate::infrastructure::Database;
use tauri::State;
use serde::{Deserialize, Serialize};

/// Order request from frontend
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlaceOrderRequest {
    pub symbol: String,
    pub side: String,
    pub order_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub price: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop_price: Option<f64>,
    pub quantity: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub client_order_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub time_in_force: Option<String>,
}

/// Convert frontend request to internal OrderRequest
impl TryFrom<PlaceOrderRequest> for OrderRequest {
    type Error = String;

    fn try_from(req: PlaceOrderRequest) -> Result<Self, Self::Error> {
        let side = req.side.parse().map_err(|e| format!("Invalid side: {}", e))?;
        let order_type = req.order_type.parse().map_err(|e| format!("Invalid order type: {}", e))?;

        Ok(OrderRequest {
            symbol: req.symbol,
            side,
            order_type,
            price: req.price,
            stop_price: req.stop_price,
            quantity: req.quantity,
            client_order_id: req.client_order_id,
            time_in_force: req.time_in_force.and_then(|t| t.parse().ok()),
        })
    }
}

/// Place a new order
#[tauri::command]
pub async fn trade_place_order(
    db: State<'_, Database>,
    user_id: String,
    request: PlaceOrderRequest,
) -> Result<ApiResponse<Order>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!("[{}] trade_place_order called: user_id={}", request_id, user_id);

    // Convert request to internal format
    let order_request: OrderRequest = request.try_into()
        .map_err(|e| format!("Invalid request: {}", e))?;

    // Get TradeService and place the order
    let trade_service = db.get_trade_service().await;
    match trade_service.place_order(order_request, &user_id).await {
        Ok(order) => {
            log::info!("[{}] Order placed successfully: {}", request_id, order.id);
            Ok(ApiResponse::success(order).with_request_id(request_id))
        }
        Err(e) => {
            log::error!("[{}] Failed to place order: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed(format!("下单失败: {}", e))).with_request_id(request_id))
        }
    }
}

/// Cancel an existing order
#[tauri::command]
pub async fn trade_cancel_order(
    db: State<'_, Database>,
    user_id: String,
    order_id: String,
) -> Result<ApiResponse<()>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!("[{}] trade_cancel_order called: user_id={}, order_id={}", request_id, user_id, order_id);

    let trade_service = db.get_trade_service().await;
    match trade_service.cancel_order(&order_id, &user_id).await {
        Ok(()) => {
            log::info!("[{}] Order canceled successfully: {}", request_id, order_id);
            Ok(ApiResponse::success_empty().with_request_id(request_id))
        }
        Err(e) => {
            log::error!("[{}] Failed to cancel order: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed(format!("撤单失败: {}", e))).with_request_id(request_id))
        }
    }
}

/// Get order by ID
#[tauri::command]
pub async fn trade_get_order(
    db: State<'_, Database>,
    user_id: String,
    order_id: String,
) -> Result<ApiResponse<Order>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!("[{}] trade_get_order called: user_id={}, order_id={}", request_id, user_id, order_id);

    let trade_service = db.get_trade_service().await;
    match trade_service.get_order(&order_id, &user_id).await {
        Ok(order) => Ok(ApiResponse::success(order).with_request_id(request_id)),
        Err(e) => {
            log::error!("[{}] Failed to get order: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::not_found("订单")).with_request_id(request_id))
        }
    }
}

/// Get orders with optional filters
#[tauri::command]
pub async fn trade_get_orders(
    db: State<'_, Database>,
    user_id: String,
    symbol: Option<String>,
    status: Option<String>,
    limit: Option<usize>,
) -> Result<ApiResponse<Vec<Order>>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    let limit = limit.unwrap_or(100);
    log::info!(
        "[{}] trade_get_orders called: user_id={}, symbol={:?}, status={:?}, limit={}",
        request_id, user_id, symbol, status, limit
    );

    let trade_service = db.get_trade_service().await;
    let order_status = status.as_ref().and_then(|s| s.parse().ok());
    match trade_service.get_orders(&user_id, symbol.as_deref(), order_status, limit).await {
        Ok(orders) => Ok(ApiResponse::success(orders).with_request_id(request_id)),
        Err(e) => {
            log::error!("[{}] Failed to get orders: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed("查询订单失败")).with_request_id(request_id))
        }
    }
}

/// Get all open orders
#[tauri::command]
pub async fn trade_get_open_orders(
    db: State<'_, Database>,
    user_id: String,
) -> Result<ApiResponse<Vec<Order>>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!("[{}] trade_get_open_orders called: user_id={}", request_id, user_id);

    let trade_service = db.get_trade_service().await;
    match trade_service.get_open_orders(&user_id).await {
        Ok(orders) => Ok(ApiResponse::success(orders).with_request_id(request_id)),
        Err(e) => {
            log::error!("[{}] Failed to get open orders: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed("查询挂单失败")).with_request_id(request_id))
        }
    }
}

/// Sync order status from exchange
#[tauri::command]
pub async fn trade_sync_order_status(
    db: State<'_, Database>,
    user_id: String,
    order_id: String,
) -> Result<ApiResponse<Order>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!(
        "[{}] trade_sync_order_status called: user_id={}, order_id={}",
        request_id, user_id, order_id
    );

    let trade_service = db.get_trade_service().await;
    match trade_service.sync_order_status(&order_id, &user_id).await {
        Ok(order) => Ok(ApiResponse::success(order).with_request_id(request_id)),
        Err(e) => {
            log::error!("[{}] Failed to sync order: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed("同步订单失败")).with_request_id(request_id))
        }
    }
}

/// Get current positions
#[tauri::command]
pub async fn trade_get_positions(
    db: State<'_, Database>,
    user_id: String,
) -> Result<ApiResponse<Vec<Position>>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!("[{}] trade_get_positions called: user_id={}", request_id, user_id);

    let trade_service = db.get_trade_service().await;
    match trade_service.get_positions(&user_id).await {
        Ok(positions) => Ok(ApiResponse::success(positions).with_request_id(request_id)),
        Err(e) => {
            log::error!("[{}] Failed to get positions: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed("查询持仓失败")).with_request_id(request_id))
        }
    }
}

/// Get account balance
#[tauri::command]
pub async fn trade_get_balance(
    db: State<'_, Database>,
) -> Result<ApiResponse<Vec<Balance>>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!("[{}] trade_get_balance called", request_id);

    let trade_service = db.get_trade_service().await;
    match trade_service.get_balance().await {
        Ok(balance) => Ok(ApiResponse::success(balance).with_request_id(request_id)),
        Err(e) => {
            log::error!("[{}] Failed to get balance: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed("查询余额失败")).with_request_id(request_id))
        }
    }
}

/// Cancel all open orders
#[tauri::command]
pub async fn trade_cancel_all_orders(
    db: State<'_, Database>,
    user_id: String,
    symbol: Option<String>,
) -> Result<ApiResponse<usize>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!(
        "[{}] trade_cancel_all_orders called: user_id={}, symbol={:?}",
        request_id, user_id, symbol
    );

    let trade_service = db.get_trade_service().await;

    // Get all open orders and cancel them one by one
    let open_orders = match trade_service.get_open_orders(&user_id).await {
        Ok(orders) => orders,
        Err(e) => {
            log::error!("[{}] Failed to get open orders: {}", request_id, e);
            return Ok(ApiResponse::error(ApiError::operation_failed("查询挂单失败")).with_request_id(request_id));
        }
    };

    let mut canceled_count = 0;
    for order in open_orders {
        if symbol.as_ref().is_none_or(|s| order.symbol == *s)
            && trade_service.cancel_order(&order.id, &user_id).await.is_ok()
        {
            canceled_count += 1;
        }
    }

    log::info!("[{}] Canceled {} orders", request_id, canceled_count);
    Ok(ApiResponse::success(canceled_count).with_request_id(request_id))
}

/// Close a position
#[tauri::command]
pub async fn trade_close_position(
    db: State<'_, Database>,
    user_id: String,
    symbol: String,
    side: String,
    quantity: Option<f64>,
) -> Result<ApiResponse<f64>, String> {
    let request_id = uuid::Uuid::new_v4().to_string();
    log::info!(
        "[{}] trade_close_position called: user_id={}, symbol={}, side={}, quantity={:?}",
        request_id, user_id, symbol, side, quantity
    );

    let trade_service = db.get_trade_service().await;

    // Get current positions
    let positions = match trade_service.get_positions(&user_id).await {
        Ok(pos) => pos,
        Err(e) => {
            log::error!("[{}] Failed to get positions: {}", request_id, e);
            return Ok(ApiResponse::error(ApiError::operation_failed("查询持仓失败")).with_request_id(request_id));
        }
    };

    // Find the position to close
    let position = match positions
        .iter()
        .find(|p| p.symbol == symbol && p.side.to_string().to_lowercase() == side.to_lowercase())
    {
        Some(p) => p,
        None => {
            return Ok(ApiResponse::error(ApiError::not_found("持仓")).with_request_id(request_id));
        }
    };

    // Determine quantity to close
    let close_qty = quantity.unwrap_or(position.quantity);

    if close_qty > position.quantity {
        return Ok(ApiResponse::error(ApiError::invalid_parameter("quantity")).with_request_id(request_id));
    }

    // Create opposite side order to close position
    let close_side_str = match position.side.to_lowercase().as_str() {
        "buy" => "sell",
        "sell" => "buy",
        _ => return Ok(ApiResponse::error(ApiError::invalid_parameter("side")).with_request_id(request_id)),
    };

    let order_side: crate::core::trade::types::OrderSide = match close_side_str.parse() {
        Ok(s) => s,
        Err(e) => return Ok(ApiResponse::error(ApiError::invalid_parameter(format!("Invalid side: {}", e))).with_request_id(request_id)),
    };

    let order_request = crate::core::trade::types::OrderRequest {
        symbol: symbol.clone(),
        side: order_side,
        order_type: crate::core::trade::types::OrderType::Market,
        price: None,
        stop_price: None,
        quantity: close_qty,
        client_order_id: Some(format!("close_{}", uuid::Uuid::new_v4())),
        time_in_force: None,
    };

    // Place the closing order
    match trade_service.place_order(order_request, &user_id).await {
        Ok(_) => {
            log::info!("[{}] Position closed successfully", request_id);
            Ok(ApiResponse::success(0.0).with_request_id(request_id))
        }
        Err(e) => {
            log::error!("[{}] Failed to close position: {}", request_id, e);
            Ok(ApiResponse::error(ApiError::operation_failed("平仓失败")).with_request_id(request_id))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_place_order_request_conversion() {
        let req = PlaceOrderRequest {
            symbol: "BTCUSDT".to_string(),
            side: "buy".to_string(),
            order_type: "market".to_string(),
            price: None,
            stop_price: None,
            quantity: 1.0,
            client_order_id: None,
            time_in_force: None,
        };

        let order_request: OrderRequest = req.try_into().unwrap();
        assert_eq!(order_request.symbol, "BTCUSDT");
        assert_eq!(order_request.side, OrderSide::Buy);
        assert_eq!(order_request.order_type, OrderType::Market);
    }

    #[test]
    fn test_place_order_request_with_limit() {
        let req = PlaceOrderRequest {
            symbol: "BTCUSDT".to_string(),
            side: "sell".to_string(),
            order_type: "limit".to_string(),
            price: Some(50000.0),
            stop_price: None,
            quantity: 0.5,
            client_order_id: Some("client123".to_string()),
            time_in_force: Some("GTC".to_string()),
        };

        let order_request: OrderRequest = req.try_into().unwrap();
        assert_eq!(order_request.side, OrderSide::Sell);
        assert_eq!(order_request.order_type, OrderType::Limit);
        assert_eq!(order_request.price, Some(50000.0));
        assert_eq!(order_request.time_in_force, Some(TimeInForce::GTC));
    }

    #[test]
    fn test_invalid_side() {
        let req = PlaceOrderRequest {
            symbol: "BTCUSDT".to_string(),
            side: "invalid".to_string(),
            order_type: "market".to_string(),
            price: None,
            stop_price: None,
            quantity: 1.0,
            client_order_id: None,
            time_in_force: None,
        };

        let result: Result<OrderRequest, _> = req.try_into();
        assert!(result.is_err());
    }

    #[test]
    fn test_invalid_order_type() {
        let req = PlaceOrderRequest {
            symbol: "BTCUSDT".to_string(),
            side: "buy".to_string(),
            order_type: "invalid".to_string(),
            price: None,
            stop_price: None,
            quantity: 1.0,
            client_order_id: None,
            time_in_force: None,
        };

        let result: Result<OrderRequest, _> = req.try_into();
        assert!(result.is_err());
    }
}
